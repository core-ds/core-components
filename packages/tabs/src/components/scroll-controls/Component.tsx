import React, { forwardRef, RefObject, useEffect, useState } from 'react';
import cn from 'classnames';
import _debounce from 'lodash.debounce';

import { IconButton } from '@alfalab/core-components-icon-button';
import { ChevronLeftCompactSIcon } from '@alfalab/icons-glyph/ChevronLeftCompactSIcon';
import { ChevronRightCompactSIcon } from '@alfalab/icons-glyph/ChevronRightCompactSIcon';

import { TabsProps } from '../../typings';

import { getDisabledState, scrollIntoFirstTab, scrollIntoLastTab } from './utils';

import styles from './index.module.css';

type ScrollControlsProps = {
    className?: string;
    view: Exclude<TabsProps['view'], undefined>;
    size: TabsProps['size'];
    containerRef: RefObject<HTMLDivElement>;
};

export const ScrollControls = forwardRef<HTMLDivElement, ScrollControlsProps>(
    ({ containerRef, view, size: sizeProp, className }, ref) => {
        const container = containerRef.current;
        const [disabledState, updateDisabledState] = useState(() => getDisabledState(container));

        useEffect(() => {
            const handleScroll = _debounce(
                () => updateDisabledState(getDisabledState(container)),
                100,
                { leading: true, maxWait: 100, trailing: true },
            );

            container?.addEventListener('scroll', handleScroll);

            return () => container?.removeEventListener('scroll', handleScroll);
        }, [container]);

        const getSize = () => {
            if (view === 'primary') {
                return sizeProp === 'xl' ? 'xs' : 'xxs';
            }

            return sizeProp && ['s', 'm', 'l', 'xl'].includes(sizeProp) ? 's' : 'xs';
        };

        const handleScrollLeft = () => scrollIntoFirstTab(container);

        const handleScrollRight = () => scrollIntoLastTab(container);

        const commonButtonProps = {
            className: styles.button,
            size: getSize(),
            view: 'secondary',
        } as const;

        return (
            <div
                ref={ref}
                className={cn(
                    styles.component,
                    styles[view],
                    sizeProp && styles[sizeProp],
                    className,
                    {
                        [styles.borderVisible]: !disabledState.toRight,
                    },
                )}
            >
                <IconButton
                    {...commonButtonProps}
                    icon={ChevronLeftCompactSIcon}
                    disabled={disabledState.toLeft}
                    onClick={handleScrollLeft}
                />
                <IconButton
                    {...commonButtonProps}
                    icon={ChevronRightCompactSIcon}
                    disabled={disabledState.toRight}
                    onClick={handleScrollRight}
                />
            </div>
        );
    },
);
