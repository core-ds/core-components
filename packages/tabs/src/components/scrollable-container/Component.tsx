import React, { ReactNode, useEffect } from 'react';
import cn from 'classnames';
import computeScrollIntoView from 'compute-scroll-into-view';

import { TabsProps } from '../../typings';

import styles from './index.module.css';

/**
 * Дополнительная прокрутка при клике на не поместившийся таб
 */
const ADDITIONAL_SCROLLLEFT_VALUE = 40;

export type ScrollableContainerProps = {
    /**
     * Дополнительный класс контейнера
     */
    containerClassName?: string;

    /**
     * Дочерние компоненты
     */
    children: ReactNode;

    /**
     * Активный элемент (всегда будет в видимой области)
     */
    activeChild: HTMLElement | null;
};

export const ScrollableContainer = ({
    containerClassName,
    children,
    activeChild,
    fullWidthScroll,
}: ScrollableContainerProps & Pick<TabsProps, 'fullWidthScroll'>) => {
    useEffect(() => {
        if (activeChild) {
            const actions = computeScrollIntoView(activeChild, {
                scrollMode: 'if-needed',
                block: 'nearest',
                inline: 'nearest',
            });

            // TODO: animate?
            actions.forEach(({ el, left }) => {
                // eslint-disable-next-line no-param-reassign
                el.scrollLeft =
                    el.scrollLeft > left
                        ? left - ADDITIONAL_SCROLLLEFT_VALUE
                        : left + ADDITIONAL_SCROLLLEFT_VALUE;
            });
        }
    }, [activeChild]);

    return (
        <div
            className={cn(styles.container, containerClassName, {
                [styles.fullWidthScroll]: fullWidthScroll,
            })}
        >
            {children}
        </div>
    );
};
