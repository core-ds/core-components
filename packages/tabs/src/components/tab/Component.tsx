import React, { forwardRef } from 'react';
import cn from 'classnames';

import { TabProps } from '../../typings';

import styles from './index.module.css';

export const Tab = forwardRef<HTMLDivElement, TabProps>(
    ({ children, hidden, className, disabled, dataTestId }, ref) =>
        children ? (
            <div
                ref={ref}
                className={cn(
                    styles.component,
                    {
                        [styles.hidden]: hidden,
                    },
                    className,
                )}
                hidden={hidden}
                role='tabpanel'
                tabIndex={disabled ? -1 : 0}
                data-test-id={dataTestId}
            >
                {children}
            </div>
        ) : null,
);
