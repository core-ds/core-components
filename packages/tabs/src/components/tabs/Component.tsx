import React, { cloneElement } from 'react';

import { TabsProps } from '../../typings';

export const Tabs = ({
    TabList,
    className,
    containerClassName,
    size,
    defaultMatch,
    children,
    selectedId,
    scrollable,
    collapsible,
    fullWidthScroll = false,
    keepMounted = false,
    dataTestId,
    onChange,
}: Omit<TabsProps, 'view'>) => {
    const tabsArray = React.Children.toArray(children) as TabsProps['children'];
    const titles = tabsArray.map(
        ({ props: { title, id, rightAddons, disabled, hidden, toggleClassName } }) => ({
            title,
            id,
            disabled,
            rightAddons,
            hidden,
            toggleClassName,
        }),
    );

    const tabs = tabsArray.filter(
        (tab) => tab.props.id === selectedId || tab.props.keepMounted || keepMounted,
    );

    return (
        <div className={className}>
            <TabList
                containerClassName={containerClassName}
                size={size}
                titles={titles}
                selectedId={selectedId}
                scrollable={scrollable}
                collapsible={collapsible}
                onChange={onChange}
                dataTestId={dataTestId}
                defaultMatch={defaultMatch}
                fullWidthScroll={fullWidthScroll}
            />

            {tabs.map((tab) => cloneElement(tab, { hidden: tab.props.id !== selectedId }))}
        </div>
    );
};
