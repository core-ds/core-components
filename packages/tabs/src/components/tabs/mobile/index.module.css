.wrapper {
    margin: 0 var(--gap-m-neg);
}
