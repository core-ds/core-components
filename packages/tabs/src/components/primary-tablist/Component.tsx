import React, { useEffect, useRef } from 'react';
import { ResizeObserver as ResizeObserverPolyfill } from '@juggle/resize-observer';
import cn from 'classnames';

import { KeyboardFocusable } from '@alfalab/core-components-keyboard-focusable';
import { fnUtils } from '@alfalab/core-components-shared';

import { useTabs } from '../../hooks/use-tabs';
import { PlatformProps, Styles, TabListProps } from '../../typings';
import { ScrollableContainer } from '../scrollable-container';
import { Title } from '../title';

export const PrimaryTabList = ({
    size,
    styles = {},
    className,
    containerClassName,
    titles = [],
    selectedId = titles.length ? titles[0].id : undefined,
    scrollable = true,
    fullWidthScroll,
    onChange,
    dataTestId,
    platform,
}: TabListProps & Styles & PlatformProps) => {
    const lineRef = useRef<HTMLDivElement>(null);

    const { selectedTab, focusedTab, getTabListItemProps } = useTabs({
        titles,
        selectedId,
        onChange,
    });

    useEffect(() => {
        if (selectedTab) {
            const updateLineWidth = () => {
                if (lineRef.current) {
                    lineRef.current.style.width = `${selectedTab.offsetWidth}px`;
                    lineRef.current.style.transform = `translateX(${selectedTab.offsetLeft}px)`;
                }
            };

            const ResizeObserver = window.ResizeObserver || ResizeObserverPolyfill;
            const observer = new ResizeObserver(updateLineWidth);

            observer.observe(selectedTab);

            return () => observer.disconnect();
        }

        return fnUtils.noop;
    }, [selectedTab]);

    const renderContent = () => (
        <div
            role='tablist'
            data-test-id={dataTestId}
            className={cn(styles.component, className, size && styles[size], {
                [styles.fullWidthScroll]: fullWidthScroll,
            })}
        >
            {titles.map(({ dataTestId: _, ref: __, ...restTitleProps }, index) => (
                <KeyboardFocusable key={restTitleProps.id}>
                    {(ref, focused) => (
                        <Title
                            {...getTabListItemProps(index, ref)}
                            {...restTitleProps}
                            focused={focused}
                            styles={styles}
                        />
                    )}
                </KeyboardFocusable>
            ))}

            <div className={styles.line} ref={lineRef} />
        </div>
    );

    return scrollable ? (
        <ScrollableContainer
            activeChild={focusedTab || selectedTab}
            containerClassName={containerClassName}
            fullWidthScroll={fullWidthScroll}
            view='primary'
            size={size}
            platform={platform}
        >
            {renderContent()}
        </ScrollableContainer>
    ) : (
        <div className={cn(styles.container, containerClassName)}>{renderContent()}</div>
    );
};
