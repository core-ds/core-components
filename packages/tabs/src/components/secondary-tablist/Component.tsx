import React from 'react';
import cn from 'classnames';

import { Tag } from '@alfalab/core-components-tag';

import { SecondaryTabListProps, Styles } from '../../typings';
import { useTabs } from '../../hooks/use-tabs';
import { ScrollableContainer } from '../scrollable-container';

export const SecondaryTabList = ({
    styles = {},
    className,
    containerClassName,
    size,
    titles = [],
    selectedId = titles.length ? titles[0].id : undefined,
    scrollable = true,
    fullWidthScroll,
    tagSize = 'xs',
    onChange,
    dataTestId,
}: SecondaryTabListProps & Styles) => {
    const { focusedTab, selectedTab, getTabListItemProps } = useTabs({
        titles,
        selectedId,
        onChange,
    });

    const renderContent = () => (
        <div
            role='tablist'
            data-test-id={dataTestId}
            className={cn(styles.component, className, size && styles[size], {
                [styles.fullWidthScroll]: fullWidthScroll,
            })}
        >
            {titles.map((item, index) => {
                if (item.hidden) return null;

                return (
                    <Tag
                        {...getTabListItemProps(index)}
                        key={item.id}
                        className={cn(styles.title, item.toggleClassName)}
                        checked={item.id === selectedId}
                        size={tagSize}
                        rightAddons={item.rightAddons}
                    >
                        {item.title}
                    </Tag>
                );
            })}
        </div>
    );

    return scrollable ? (
        <ScrollableContainer
            activeChild={focusedTab || selectedTab}
            containerClassName={containerClassName}
            fullWidthScroll={fullWidthScroll}
        >
            {renderContent()}
        </ScrollableContainer>
    ) : (
        <div className={cn(styles.container, containerClassName)}>{renderContent()}</div>
    );
};
