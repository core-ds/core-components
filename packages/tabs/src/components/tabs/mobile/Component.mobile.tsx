import React from 'react';

import { TabsProps } from '../../../typings';
import { PrimaryTabListMobile } from '../../primary-tablist/Component.mobile';
import { SecondaryTabListMobile } from '../../secondary-tablist/Component.mobile';
import { Tabs } from '../Component';
import styles from './index.module.css';

const views = {
    primary: PrimaryTabListMobile,
    secondary: SecondaryTabListMobile,
};

export type TabsMobileProps = Omit<TabsProps, 'TabList' | 'size'> & {
    /**
     * Игнорировать паддинги страницы в 16px
     */
    fullWidth?: boolean;
};

export const TabsMobile = ({
    view = 'primary',
    scrollable = true,
    fullWidth = false,
    ...restProps
}: TabsMobileProps) => {
    const Component = <Tabs TabList={views[view]} scrollable={scrollable} {...restProps} />;

    return fullWidth ? <div className={styles.wrapper}>{Component}</div> : Component;
};
