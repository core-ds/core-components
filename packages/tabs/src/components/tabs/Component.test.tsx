/* eslint-disable multiline-comment-style */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { TabsDesktop } from './Component.desktop';
import { TabsMobile } from './Component.mobile';
import { TabsResponsive } from './Component.responsive';
import { Tab } from '../tab';
import { TabsProps } from '../../typings';
import { DiamondsMIcon } from '@alfalab/icons-glyph/DiamondsMIcon';

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => {
        return {
            matches: true,
            media: query,
            addListener: jest.fn(),
            removeListener: jest.fn(),
        };
    }),
});

const tabVariants: Array<
    [typeof TabsMobile | typeof TabsDesktop | typeof TabsResponsive, TabsProps['view']]
> = [
    [TabsDesktop, 'primary'],
    [TabsDesktop, 'secondary'],
    [TabsMobile, 'primary'],
    [TabsMobile, 'secondary'],
    [TabsResponsive, 'primary'],
    [TabsResponsive, 'secondary'],
];

const renderTabs = (
    Component: typeof TabsMobile | typeof TabsDesktop | typeof TabsResponsive,
    props = {},
    tabProps = {},
) =>
    render(
        <Component selectedId='tab-2' {...props}>
            <Tab title='Таб 1' id='tab-1' rightAddons='addon' {...tabProps}>
                Таб 1
            </Tab>
            <Tab title='Таб 2' id='tab-2' {...tabProps}>
                Таб 2
            </Tab>
            <Tab title='Таб 3' id='tab-3' {...tabProps}>
                Таб 3
            </Tab>
            <Tab title='Таб 4' id='tab-4' {...tabProps}>
                Таб 4
            </Tab>
            <Tab title='Таб 5' id='tab-5' {...tabProps}>
                Таб 5
            </Tab>
        </Component>,
    );

describe('Tabs', () => {
    describe('Snapshots tests', () => {
        it.each(tabVariants)('should match snapshot', (Component, view) => {
            const { container } = renderTabs(Component, { view });
            expect(container).toMatchSnapshot();
        });
    });

    it('should match snapshot with icon as title', () => {
        const { container } = render(
            <TabsDesktop selectedId='tab-2'>
                <Tab title={<DiamondsMIcon />} id='tab-1' />
                <Tab title={<DiamondsMIcon />} id='tab-2' />
            </TabsDesktop>,
        );
        expect(container).toMatchSnapshot();
    });

    describe('Classes tests', () => {
        it.each(tabVariants)('should set custom class', (Component, view) => {
            const className = 'custom-class';
            const { container } = renderTabs(Component, { view, className });

            expect(container.firstElementChild).toHaveClass(className);
        });

        it.each([tabVariants[0]])('should set custom container class', (Component, view) => {
            const containerClassName = 'custom-container-class';
            const { getByTestId } = renderTabs(Component, {
                view,
                containerClassName,
                dataTestId: 'tabs-test',
            });
            const container = getByTestId('tabs-test').parentElement as HTMLElement;
            expect(container).toHaveClass(containerClassName);
        });
    });

    describe('Attributes tests', () => {
        it.each(tabVariants)('should set `data-test-id` attribute', (Component, view) => {
            const dataTestId = 'test-id';
            const toggleTestId = 'tab';
            const { getByTestId, getAllByTestId } = renderTabs(
                Component,
                { view, dataTestId },
                { dataTestId: 'tab' },
            );

            const tabs = getAllByTestId(toggleTestId + '-toggle');

            expect(getByTestId(dataTestId).getAttribute('role')).toBe('tablist');
            expect(tabs[1].textContent).toBe('Таб 2');
            expect(tabs[2].textContent).toBe('Таб 3');
        });

        it.each(tabVariants)('should set custom style', (Component, view) => {
            const style = { padding: 20 };
            const { container } = renderTabs(Component, { view, style });

            const firstElement = container.firstChild;

            expect(firstElement).toHaveStyle('padding: 20px');
        });
    });

    describe('Render tests', () => {
        it.each(tabVariants)(
            'should render only selected tabpanel by default',
            (Component, view) => {
                const { getAllByRole } = renderTabs(Component, { view });
                const tabpanels = getAllByRole('tabpanel');

                expect(tabpanels.length).toEqual(1);
            },
        );

        it.each(tabVariants)(
            'should render only one visible tabpanel if keepMounted=`true`',
            (Component, view) => {
                const { getAllByRole } = renderTabs(Component, { view, keepMounted: true });
                const hidden = getAllByRole('tabpanel', { hidden: true });
                const visible = getAllByRole('tabpanel', { hidden: false });

                expect(hidden.length).toBeGreaterThan(1);
                expect(visible.length).toEqual(1);
            },
        );

        it.each(tabVariants)('should render corresponding tabpanel', (Component, view) => {
            const { getByRole } = renderTabs(Component, { view });
            const selectedTab = getByRole('tab', { selected: true });
            const tabpanel = getByRole('tabpanel', { hidden: false });

            expect(tabpanel.textContent).toEqual(selectedTab.textContent);
        });

        it.each(tabVariants)('should unmount without errors', (Component, view) => {
            const { unmount } = renderTabs(Component, { view });

            expect(unmount).not.toThrowError();
        });

        it('should not render empty tabs', () => {
            const { queryAllByRole } = render(
                <TabsDesktop selectedId='tab-2' keepMounted={true}>
                    <Tab title='Таб 1' id='tab-1' />
                    <Tab title='Таб 2' id='tab-2' />
                    <Tab title='Таб 3' id='tab-3'>
                        Таб 3
                    </Tab>
                </TabsDesktop>,
            );

            expect(queryAllByRole('tabpanel', { hidden: true })).toHaveLength(1);
            expect(queryAllByRole('tabpanel', { hidden: false })).toHaveLength(0);
        });
    });

    describe('Interaction tests', () => {
        it.each(tabVariants)('should call `onChange`', (Component, view) => {
            const cb = jest.fn();
            const { getAllByRole } = renderTabs(Component, { view, onChange: cb });
            const tab = getAllByRole('tab', { selected: false });

            if (tab.length > 0) {
                fireEvent.click(tab[0]);
            }

            expect(cb).toBeCalledTimes(1);
        });

        it.each(tabVariants)('should not call `onChange` for selected tab', (Component, view) => {
            const cb = jest.fn();
            const { getByRole } = renderTabs(Component, { view, onChange: cb });
            const tab = getByRole('tab', { selected: true });

            if (tab) {
                fireEvent.click(tab);
            }

            expect(cb).not.toBeCalled();
        });
    });
});
