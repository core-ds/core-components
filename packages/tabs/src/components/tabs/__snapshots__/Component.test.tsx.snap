// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Snapshots tests should match snapshot 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="component m"
        role="tablist"
      >
        <button
          aria-selected="false"
          class="title"
          id="tab-1"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 1
          </span>
          <span
            class="rightAddons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="title selected"
          id="tab-2"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="content"
          >
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-4"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-5"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 5
          </span>
        </button>
        <div
          class="line"
          style="width: 0px; transform: translateX(0px);"
        />
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 2`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="component s"
        role="tablist"
      >
        <button
          aria-selected="false"
          class="component component s defaultVariant withRightAddons title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 1
          </span>
          <span
            class="addons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="component component s defaultVariant checked title"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 5
          </span>
        </button>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 3`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class=""
        role="tablist"
      >
        <button
          aria-selected="false"
          class=""
          id="tab-1"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 1
          </span>
          <span>
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class=""
          id="tab-2"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class=""
          >
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class=""
          id="tab-3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class=""
          id="tab-4"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class=""
          id="tab-5"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 5
          </span>
        </button>
        <div
          style="width: 0px; transform: translateX(0px);"
        />
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 4`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class=""
        role="tablist"
      >
        <button
          aria-selected="false"
          class="component component xs defaultVariant withRightAddons"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 1
          </span>
          <span
            class="addons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="component component xs defaultVariant checked"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component xs defaultVariant"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component xs defaultVariant"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component xs defaultVariant"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 5
          </span>
        </button>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 5`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="component m"
        role="tablist"
      >
        <button
          aria-selected="false"
          class="title"
          id="tab-1"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 1
          </span>
          <span
            class="rightAddons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="title selected"
          id="tab-2"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="content"
          >
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-4"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          id="tab-5"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="content"
          >
            Таб 5
          </span>
        </button>
        <div
          class="line"
          style="width: 0px; transform: translateX(0px);"
        />
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 6`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="component s"
        role="tablist"
      >
        <button
          aria-selected="false"
          class="component component s defaultVariant withRightAddons title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 1
          </span>
          <span
            class="addons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="component component s defaultVariant checked title"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="component component s defaultVariant title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 5
          </span>
        </button>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;
